File: problems_001to025/problem_023/solution.py
import os
import sys

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch
from toolbox import natural_sum, positive_divisors


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = None
        abundant_numbers = []
        two_abundant_sums = set()
        for num in range(1, 28124):
            proper_divisors = positive_divisors(num)[:-1]
            divisor_sum = sum(proper_divisors)
            if divisor_sum > num:
                abundant_numbers += [num]
                for abundant in abundant_numbers:
                    abundant_sum = abundant + num
                    if abundant_sum > 28123:
                        break
                    else:
                        two_abundant_sums.add(abundant_sum)
        answer = natural_sum(28123) - sum(two_abundant_sums)
        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_001to025/problem_023/solution.js
const { sum } = require('mathjs');

const { naturalSum, positiveDivisors } = require('../../lib/toolbox');

const solution = () => {
  let answer = null;

  const abundantNumbers = [];
  const twoAbundantSums = new Set();

  for (let num = 1; num <= 28123; num += 1) {
    const properDivisors = positiveDivisors(num).slice(0, -1);
    const divisorSum = sum(properDivisors);

    if (divisorSum > num) {
      abundantNumbers.push(num);

      abundantNumbers.every((abundant) => {
        const abundantSum = abundant + num;

        if (abundantSum <= 28123) {
          twoAbundantSums.add(abundantSum);
        }

        return abundantSum <= 28123;
      });
    }
  }

  answer = naturalSum(28123) - sum([...twoAbundantSums]);

  return answer;
};

if (process.env.NODE_ENV !== 'test') {
  // eslint-disable-next-line no-console
  console.log(solution());
}

module.exports = solution;
